### PR TITLE
fuzz: set/get D-Bus properties as well

### PR DESCRIPTION
**Pull request by mrc0mmand**
Resolves: #61

### CODE REVIEW

**Comment by mrc0mmand** [comment]
@evverx when you have a bit of time, could you give this one a spin? It should, hopefully, do what it's supposed to (I still have a couple of cleanup-related patches on my mind, but nothing major).

I suppose the suppression stuff should be extended (probably just an optional `m/p` prefix), so we're able to suppress properties as well.

**Comment by evverx** [comment]
It seems to be working. Thanks! It takes too long though and looks like it would make it a bit harder to integrate dfuzzer into the systemd CI. To mitigate it I think it would make sense to test every readable property 2 times (which I think should be enough to trigger most "shallow" issues like memory leaks and so on). Writable properties should be tested more thoroughly but without dictionaries it would be a waste of time mostly. Also I think it would be great if it was possible to skip properties/methods entirely with something like `--skip-properties/--skip-methods` and maybe `--skip-properties` should be turned off by default. (to test this PR I commented out the part where methods are tested).

I have no idea how it can be integrated with corpora and machine-readable logs.

> I suppose the suppression stuff should be extended (probably just an optional m/p prefix), so we're able to suppress properties as well.

I think ideally for it to be consistent it should be possible to suppress properties by analogy with methods but I have to admit I'm not sure why I would do that (at this point). I pointed that at systemd though where "writable" properties are set with the "set-transient-property" method. I can imagine scenarios where writable properties should be suppressed.

File: meson.build [hunk 1]
```diff
@@ -10,6 +10,10 @@ project('dfuzzer', 'c',
 libgio = dependency('gio-2.0', required : true)
 xsltproc = find_program('xsltproc', required: false)
 
+if get_option('b_coverage')
+        add_project_arguments('-DWITH_COVERAGE', language : 'c')
+endif
```
Comment (evverx) [inline meson.build:15]: @mrc0mmand could you move this commit to a separate PR?

Comment (mrc0mmand) [reply meson.build:15]: Sure, will do!

**Comment by mrc0mmand** [comment]
@evverx I tweaked the limits a bit following your suggestion and the properties are now read up to 2 times and written to up to 16 times per property, which makes it much faster. Also, I added the `--skip-properties` and `--skip-methods` options, but I left them both off by default, since the runtime even with properties now seems reasonable.

**Comment by evverx** [comment]
Thanks! I'll try to take a look tomorrow.

As far as I can see this PR broke (understandably) an undocumented feature I actively used. Without this PR I can run something like
```
dfuzzer -v -n org.freedesktop.systemd1 -t Clean
```
to go over all the object/interfaces and call the "Clean" method. With this patch applied `dfuzzer` seems to fail as soon as it starts with
```
[CONNECTED TO PID: 1]
Object: /
 Interface: org.freedesktop.DBus.Peer
Error: Method 'Clean' is not in the interface 'org.freedesktop.DBus.Peer'.
Exit status: 1
```
I wonder if it would be possible to bring that feature back and make it "official".

**Comment by evverx** [comment]
> I wonder if it would be possible to bring that feature back and make it "official".

On second thought it would probably make sense to figure out how to support use cases like that better first. I think it would be better if it was possible to pass wildcards like `-o * -i *` to make `dfuzzer` iterate over all objects and interfaces and something like `-o /org/freedesktop/systemd1/unit/` to go over units. Let's put that feature on hold. Probably I should stop using undocumented features :-)

**Comment by mrc0mmand** [comment]
> > I wonder if it would be possible to bring that feature back and make it "official".
> 
> On second thought it would probably make sense to figure out how to support use cases like that better first. I think it would be better if it was possible to pass wildcards like `-o * -i *` to make `dfuzzer` iterate over all objects and interfaces and something like `-o /org/freedesktop/systemd1/unit/` to go over units. Let's put that feature on hold. Probably I should stop using undocumented features :-)

That would make sense, I guess. Anyway, returning that feature is basically a two-line change, so that shouldn't be an issue either, I can look into that tomorrow :-)

**Review by evverx — APPROVED** [review]
LGTM. Thanks!

I think the logs need fixing because they don't seem to reflect reality
```sh
  PASS [P] InitRDUnitsLoadStartTimestamp
  Property: InitRDUnitsLoadStartTimestampMonotonic (t) => 32 iterations
  [P] InitRDUnitsLoadStartTimestampMonotonic...Got value for property org.freedesktop.systemd1.Manager.InitRDUnitsLoadStartTimestampMonotonic: (<uint64 0>,)
```
(32 should be replaced with 2 there). Plus it doesn't seem to be possible to tell "read" and read-write" properties apart with just `[P]`.

I'll go ahead and merge the PR to let Coverity Scan analyze it.

**Comment by evverx** [comment]
Now that properties are tested I think `dfuzzer` shouldn't call the methods of the `org.freedesktop.DBus.Properties` interface of every object too much. It should help to speed `dfuzzer` up a bit.

**Comment by mrc0mmand** [comment]
> LGTM. Thanks!
> 
> I think the logs need fixing because they don't seem to reflect reality
> 
> ```shell
>   PASS [P] InitRDUnitsLoadStartTimestamp
>   Property: InitRDUnitsLoadStartTimestampMonotonic (t) => 32 iterations
>   [P] InitRDUnitsLoadStartTimestampMonotonic...Got value for property org.freedesktop.systemd1.Manager.InitRDUnitsLoadStartTimestampMonotonic: (<uint64 0>,)
> ```
> 
> (32 should be replaced with 2 there). Plus it doesn't seem to be possible to tell "read" and read-write" properties apart with just `[P]`.
> 

Yeah, I noticed the logs don't match the reality in some other places as well, I'll take a look.

> Now that properties are tested I think `dfuzzer` shouldn't call the methods of the `org.freedesktop.DBus.Properties` interface of every object too much. It should help to speed `dfuzzer` up a bit.

Good point!
